packages/express/src/index.ts

```diff
@@ -0,0 +1,93 @@
+import { IncomingMessage } from 'http';
+
+import NodeClient from '@logto/node';
+import { Request, Response, NextFunction, Router } from 'express';
+
+import ExpressStorage from './storage';
+import { LogtoExpressConfig } from './types';
+
+export type Middleware = (
+  request: Request,
+  response: Response,
+  next: NextFunction
+) => Promise<void>;
+
+export default class LogtoClient {
+  constructor(private readonly config: LogtoExpressConfig) {}
+
+  handleSignIn =
+    (redirectUri = `${this.config.baseUrl}/logto/sign-in-callback`): Middleware =>
+    async (request, response) => {
+      const nodeClient = this.createNodeClient(request, response);
+      await nodeClient.signIn(redirectUri);
+    };
+
+  handleSignInCallback =
+    (redirectTo = this.config.baseUrl): Middleware =>
+    async (request, response) => {
+      const nodeClient = this.createNodeClient(request, response);
+
+      if (request.url) {
+        await nodeClient.handleSignInCallback(`${this.config.baseUrl}${request.originalUrl}`);
+        response.redirect(redirectTo);
+      }
+    };
+
+  handleSignOut =
+    (redirectUri = this.config.baseUrl): Middleware =>
+    async (request, response) => {
+      const nodeClient = this.createNodeClient(request, response);
+      await nodeClient.signOut(redirectUri);
+    };
+
+  handleAuthRoutes = (): Router => {
+    // eslint-disable-next-line new-cap
+    const router = Router();
+
+    router.use('/logto/:action', (request, response, next) => {
+      const { action } = request.params;
+
+      if (action === 'sign-in') {
+        return this.handleSignIn()(request, response, next);
+      }
+
+      if (action === 'sign-in-callback') {
+        return this.handleSignInCallback()(request, response, next);
+      }
+
+      if (action === 'sign-out') {
+        return this.handleSignOut()(request, response, next);
+      }
+
+      response.status(404).end();
+    });
+
+    return router;
+  };
+
+  private createNodeClient(request: IncomingMessage, response: Response) {
+    this.checkSession(request);
+    const storage = new ExpressStorage(request);
+
+    return new NodeClient(
+      {
+        ...this.config,
+        persistAccessToken: this.config.persistAccessToken ?? true,
+      },
+      {
+        storage,
+        navigate: (url) => {
+          response.redirect(url);
+        },
+      }
+    );
+  }
+
+  private checkSession(request: IncomingMessage) {
+    // We assume that `session` is configured in the express app, but need to check it there.
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (!request.session) {
+      throw new Error('You should configure express-session before using Logto express SDK.');
```

Comment:
Better throwing this error as a LogtoClientError with an error code, as one day we are going to i18n these error messages.

Reply:
nice, updated